src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -169,48 +141,66 @@ private class RsFnInferenceContext(
+            inferStructTypeArguments(expr)
+            return TyUnknown
+        }
+        val (element, subst) = boundElement
         return when (element) {
             is RsStructItem -> element.type
-                .substitute(subst)
-                .substitute(inferStructTypeArguments(expr))
             is RsEnumVariant -> element.parentEnum.type
-                .substitute(subst)
-                .substitute(inferStructTypeArguments(expr))
             else -> TyUnknown
-        }
+        }.substitute(subst).substitute(inferStructTypeArguments(expr))
+    }
+
+    private fun inferRsTupleExprType(expr: RsTupleExpr): Ty {
+        return TyTuple(expr.exprList.map { inferExprType(it) })
+    }
+
+    fun inferCastExprType(expr: RsCastExpr): Ty {
+        inferExprType(expr.expr)
+        return expr.typeReference.type
     }
 
     private fun inferCallExprType(expr: RsCallExpr): Ty {
-        val fn = expr.expr
-        val ty = fn.ty
+        val ty = inferExprType(expr.expr)
+        val argExprs = expr.valueArgumentList.exprList
         // `struct S; S();`
-        if (ty is TyStructOrEnumBase && expr.valueArgumentList.exprList.isEmpty()) return ty
+        if (ty is TyStructOrEnumBase && argExprs.isEmpty()) return ty
 
-        val calleeType = lookup.asTyFunction(ty) ?: return TyUnknown
-        return calleeType.retType.substitute(mapTypeParameters(calleeType.paramTypes, expr.valueArgumentList.exprList))
+        val calleeType = lookup.asTyFunction(ty) ?: return inferUnresolvedFnCallExprType(argExprs)
+        return calleeType.retType.substitute(mapTypeParameters(calleeType.paramTypes, argExprs))
     }
 
     private fun inferMethodCallExprType(methodCall: RsMethodCall): Ty {
-        val receiver = methodCall.receiver.ty
+        val receiver = inferExprType(methodCall.receiver)
+        val argExprs = methodCall.valueArgumentList.exprList
         val (method, subst) = resolveMethodCallReferenceWithReceiverType(lookup, receiver, methodCall)
-            .firstOrNull()?.downcast<RsFunction>() ?: return TyUnknown
+            .firstOrNull()?.downcast<RsFunction>() ?: return inferUnresolvedFnCallExprType(argExprs)
 
         val returnType = (method.retType?.typeReference?.type ?: TyUnit)
             .substitute(subst)
             .substitute(mapOf(TyTypeParameter.self() to receiver))
 
-        val methodType = method.type as? TyFunction ?: return returnType
+        val methodType = method.type as? TyFunction ?: return inferUnresolvedFnCallExprType(argExprs)
         // drop first element of paramTypes because it's `self` param
         // and it doesn't have value in `methodCall.valueArgumentList.exprList`
-        return returnType.substitute(mapTypeParameters(methodType.paramTypes.drop(1), methodCall.valueArgumentList.exprList))
+        return returnType.substitute(mapTypeParameters(methodType.paramTypes.drop(1), argExprs))
+    }
+
+    private fun inferUnresolvedFnCallExprType(argExprs: List<RsExpr>): Ty {
```

Comment:
Hm, are we walking arguments if function call is resolved? Looks like there's a subtle invariant here, that on every exit arguments must be walked. Let's just walk arguments unconditionally on entering the function, and `return TyUnknown` elsewhere. 

The particular smell here is that this is a function, intended primarily for side effects, but whose return value is not Unit.